enterprise/server/build_event_publisher/build_event_publisher.go

```diff
@@ -88,98 +92,153 @@ func (p *Publisher) run(ctx context.Context) {
+		if err := stream.Send(req); err != nil {
+			return status.UnavailableErrorf("send failed: %s", err)
 		}
 	}
 	// After successfully transmitting all events, close our side of the stream
 	// and wait for server ACKs before closing the connection.
 	if err := stream.CloseSend(); err != nil {
-		p.setError(status.WrapError(err, "failed to close build event stream"))
-		return
+		return status.UnavailableErrorf("close send failed: %s", err)
 	}
-	<-doneReceiving
+	// Return any error from receiving ACKs
+	return <-doneReceiving
 }
 
-func (p *Publisher) Publish(e *bespb.BuildEvent) error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	if p.err != nil {
-		return status.WrapError(p.err, "cannot publish event due to previous error")
+func (p *Publisher) Publish(bazelEvent *bespb.BuildEvent) error {
+	bazelEventAny, err := ptypes.MarshalAny(bazelEvent)
+	if err != nil {
+		return err
 	}
-	p.events <- e
+	be := &bepb.BuildEvent{
+		EventTime: ptypes.TimestampNow(),
+		Event:     &bepb.BuildEvent_BazelEvent{BazelEvent: bazelEventAny},
+	}
+	p.events.Add(be)
 	return nil
 }
 
-func (p *Publisher) Wait() error {
-	p.events <- nil
-	<-p.done
-	return p.err
+// Finish publishes the final stream event and then waits for the server to ACK
+// all published events.
+func (p *Publisher) Finish() error {
+	be := &bepb.BuildEvent{
+		EventTime: ptypes.TimestampNow(),
+		Event: &bepb.BuildEvent_ComponentStreamFinished{
+			ComponentStreamFinished: &bepb.BuildEvent_BuildComponentStreamFinished{
+				Type: bepb.BuildEvent_BuildComponentStreamFinished_FINISHED,
+			},
+		},
+	}
+	p.events.Add(be)
+	if err := <-p.done; err != nil {
+		return status.UnavailableErrorf("failed to publish build event stream after multiple attempts: %s", err)
+	}
+	return nil
+}
+
+// EventBuffer holds build events that have not been ack'd by the server.
+//
+// It supports a single consumer and multiple producers: multiple goroutines can
+// call Add(), and a single goroutine can call Subscribe() to get the currently
+// buffered events and subsequently added events.
+type EventBuffer struct {
+	streamID *bepb.StreamId
+
+	mu sync.Mutex
+	// TODO(bduffany): Store these on disk since the buffer can get quite large if
+	// there is a lot of log output.
+	items []*bepb.BuildEvent
+	// notify is used to signal that an event has been added to the buffer.
+	notify chan struct{}
+}
+
+func NewEventBuffer(streamID *bepb.StreamId) *EventBuffer {
+	return &EventBuffer{
+		streamID: streamID,
+		notify:   make(chan struct{}, 1),
+	}
+}
+
+// Subscribe returns a channel that streams the currently buffered events as
+// well as any subsequently added events, up until the ComponentStreamFinished
+// event.
+//
+// The returned callback function cancels the subscription. It must be called,
+// or else a goroutine leak may occur. Once cancel is called, it is safe for
+// another goroutine to call Subscribe() again.
+func (b *EventBuffer) Subscribe() (<-chan *pepb.OrderedBuildEvent, func()) {
```

Comment:
it feels a little strange to return a channel to read from (read) and a cancel func (write/cancel) instead of having the producer close the channel. Since you're just deferring the cancel at the Subscribe() callsite -- could we instead clean things up another way?

Reply:
The producer does close the channel (`defer close(events)` at the start of the goroutine). But the goroutine needs a way to get notified that the receiver has stopped receiving values, otherwise it will get blocked indefinitely when trying to send on the channel. I could be missing something though -- maybe there is a cleaner way to do this?